[Temporal] balanceTime() should take Int128 arguments

catamorphism opened the pull request:
#### ce3c6e45d71d35adf3278d6defeb0e27d2308d63
<pre>
[Temporal] balanceTime() should take Int128 arguments
<a href="https://bugs.webkit.org/show_bug.cgi?id=301710">https://bugs.webkit.org/show_bug.cgi?id=301710</a>

Reviewed by Yusuke Suzuki.

When adding a duration with a microseconds or nanoseconds value that is
close to Number.MAX_SAFE_INTEGER, `TemporalPlainTime::addTime()` was
doing an addition that overflowed, as the Duration and PlainTime fields
are doubles.

This exposed another bug in balanceTime() (when the modulo result is
negative, the next-biggest time unit should be decremented by 1).

This also fixes two test262 tests.
* JSTests/stress/temporal-plaindatetime.js:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/TemporalObject.cpp:
(JSC::nonNegativeModulo): Deleted.
* Source/JavaScriptCore/runtime/TemporalObject.h:
* Source/JavaScriptCore/runtime/TemporalPlainTime.cpp:
(JSC::balanceTime):
(JSC::TemporalPlainTime::addTime):

Canonical link: <a href="https://commits.webkit.org/302509@main">https://commits.webkit.org/302509@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3cc6f9d1b9d7700a61fd01ea04808d7a3242d334

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129307 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1563 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40144 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136684 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80698 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1493 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1440 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98478 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66379 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132254 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1169 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115824 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79121 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1093 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33945 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79961 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/121299 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109543 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34446 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139156 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/127759 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1352 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1311 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107005 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1396 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112160 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106839 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27205 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1112 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30681 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/53984 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1424 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64788 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/160773 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1242 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40116 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1277 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1346 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainTime.cpp:
```diff
@@ -483,10 +517,12 @@ ISO8601::Duration TemporalPlainTime::addTime(const ISO8601::PlainTime& plainTime
-        plainTime.microsecond() + duration.microseconds(),
-        plainTime.nanosecond() + duration.nanoseconds());
+// The seconds, milliseconds, microseconds, or nanoseconds fields can be up to
+// MAX_SAFE_INTEGER. So to do the addition, we have to convert to Int128.
```
Add indents here.

catamorphism replied:
Fixed in [ded265c](https://github.com/WebKit/WebKit/pull/53211/commits/ded265ca7cec5934d1a5b7b1ed4c6e9c764862b4)

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainTime.cpp:
```diff
@@ -144,23 +144,53 @@ TemporalPlainTime* TemporalPlainTime::tryCreateIfValid(JSGlobalObject* globalObj
 
 // https://tc39.es/proposal-temporal/#sec-temporal-balancetime
-static ISO8601::Duration balanceTime(double hour, double minute, double second, double millisecond, double microsecond, double nanosecond)
+static ISO8601::Duration balanceTime(double hour, double minute, Int128 second, Int128 millisecond, Int128 microsecond, Int128 nanosecond)
```
Why are hour and minute double?

catamorphism replied:
The values hours and minutes can have in a Duration are more constrained because there is a limit on the total of Duration time units ( https://tc39.es/proposal-temporal/#sec-isvalidduration step 8) and if hours or minutes exceeded the maximum safe integer, that would have been rejected because the total for the duration would exceed this limit.

But it would be perfectly correct to make hours and minutes Int128 as well.

Constellation replied:
OK, so please make it to Int128.

catamorphism replied:
Done in [5d33910](https://github.com/WebKit/WebKit/pull/53211/commits/5d33910fa25d2ff7b9145c94c00be0b8c830d2be)

Constellation reviewed (commented):
Commented

Constellation commented on Source/JavaScriptCore/runtime/TemporalPlainTime.cpp:
```diff
@@ -144,23 +144,53 @@ TemporalPlainTime* TemporalPlainTime::tryCreateIfValid(JSGlobalObject* globalObj
+        millisecond += 1000;
+    }
+
+    minute += static_cast<double>(second / 60);
```
Let's not use double.

catamorphism replied:
Sorry, missed that one! Fixed in [80b40a4](https://github.com/WebKit/WebKit/pull/53211/commits/80b40a43f19b5e3ef4a63f5638065d412154d769)

catamorphism committed:
[Temporal] balanceTime() should take Int128 arguments
https://bugs.webkit.org/show_bug.cgi?id=301710

Reviewed by Yusuke Suzuki.

When adding a duration with a microseconds or nanoseconds value that is
close to Number.MAX_SAFE_INTEGER, `TemporalPlainTime::addTime()` was
doing an addition that overflowed, as the Duration and PlainTime fields
are doubles.

This exposed another bug in balanceTime() (when the modulo result is
negative, the next-biggest time unit should be decremented by 1).

This also fixes two test262 tests.
* JSTests/stress/temporal-plaindatetime.js:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/TemporalObject.cpp:
(JSC::nonNegativeModulo): Deleted.
* Source/JavaScriptCore/runtime/TemporalObject.h:
* Source/JavaScriptCore/runtime/TemporalPlainTime.cpp:
(JSC::balanceTime):
(JSC::TemporalPlainTime::addTime):

Canonical link: https://commits.webkit.org/302509@main
